ccm_web/server/src/api/api.admin.handler.ts

```diff
@@ -103,4 +115,113 @@ export class AdminApiHandler {
     logger.debug(`Time elapsed: (${(end - start) / NS_PER_SEC}) seconds`)
     return finalResult
   }
+
+  // async createExternalUser (user: SectionExternalUserDto, accountID: number): Promise<CanvasUserLoginEmail | APIErrorData> {
+  async createExternalUser (user: SectionExternalUserDto, accountID: number): Promise<CanvasUserLoginEmail> {
+    const email = user.email
+    const loginId = email.replace('@', '+')
+    const fullName = `${user.givenName} ${user.surname}`
+    const sortableName = `${user.surname}, ${user.givenName}`
+
+    try {
+      const endpoint = `accounts/${accountID}/users`
+      const method = HttpMethod.Post
+      const body: any = {
+        user: {
+          // API doesn't provide separate given- and surname fields
+          name: fullName,
+          sortable_name: sortableName,
+          skip_registration: true
+        },
+        pseudonym: {
+          unique_id: loginId,
+          send_confirmation: false
+        },
+        communication_channel: {
+          type: 'email',
+          address: email,
+          skip_confirmation: true
+        },
+        force_validations: false
+      }
+      logger.debug(`Sending admin request to Canvas endpoint: "${endpoint}"; method: "${method}"; body: "${JSON.stringify(body)}"`)
+      const response = await this.requestor.request<CanvasUserLoginEmail>(endpoint, method, body)
+      logger.debug(`Received response with status code (${String(response.statusCode)})`)
+      const {
+        id,
+        name,
+        sortable_name, // eslint-disable-line
+        short_name, // eslint-disable-line
+        login_id // eslint-disable-line
+      } = response.body
+      return {
+        id,
+        name,
+        sortable_name,
+        short_name,
+        login_id,
+        email
+      }
+    } catch (error: any) {
+      // const errorResponse = handleAPIError(error, `Login ID: ${loginId}; Role: ${user.type}`)
+      // return {
+      //   statusCode: errorResponse.canvasStatusCode,
+      //   errors: [errorResponse]
+      // }
+      return {
+        name: fullName,
+        sortable_name: sortableName,
+        login_id: loginId,
+        email: email,
+        error: {
+          statusCode: error.response.statusCode,
+          body: error.response.body
+        }
+      }
```

Comment:
Okay, I see the issue now. Because the requests are not sequential, we need a way to tie the errors to the user's. We could use `failedInput` and just pass the `loginId`, but that is a bit clunky...

Reply:
This was changed as part of the changes from my PR (https://github.com/lsloan/ccm/pull/5).